snuba/query/snql/parser.py

```diff
@@ -108,31 +167,52 @@ class OrTuple(NamedTuple):
     exp: Expression
 
 
+class EntityTuple(NamedTuple):
+    alias: str
+    name: str
+    sample_rate: Optional[float]
+
+
+class RelationshipTuple(NamedTuple):
+    lhs: EntityTuple
+    relationship: str
+    rhs: EntityTuple
+
+
 class SnQLVisitor(NodeVisitor):
     """
     Builds Snuba AST expressions from the Parsimonious parse tree.
     """
 
     def visit_query_exp(self, node: Node, visited_children: Iterable[Any]) -> Query:
-        _, where, collect, groupby, having, orderby = visited_children
-        # check for empty clauses
-        if isinstance(groupby, Node):
-            groupby = None
-        if isinstance(orderby, Node):
-            orderby = None
-        if isinstance(having, Node):
-            having = None
-        return Query(
-            body={},
-            data_source=None,
-            selected_columns=collect,
-            array_join=None,
-            condition=where,
-            prewhere=None,
-            groupby=groupby,
-            having=having,
-            order_by=orderby,
-        )
+        args: MutableMapping[str, Any] = {
+            "body": {},
+            "array_join": None,
+            "prewhere": None,
+        }
+        (
+            args["data_source"],
+            args["selected_columns"],
+            args["groupby"],
+            args["condition"],
+            args["having"],
+            args["order_by"],
+            args["limitby"],
+            args["limit"],
+            offset,
+            args["sample"],
+        ) = visited_children
+        if not isinstance(offset, Node):
+            args["offset"] = offset
+
+        for k, v in args.items():
+            if isinstance(v, Node):
+                args[k] = None
+
+        if args["data_source"].sample_rate is not None:
+            args["sample"] = args["data_source"].sample_rate
+
+        return Query(**args)
```

Comment:
Does mypy actually like this ? It should think those are all `Any` since visited_children is an Iterable of Any